### ISSUE TITLE
`ltriang` fails on a wide matrix.

### ISSUE THREAD

**Issue by hurak** [neveritt/PolynomialMatrices.jl#37]
A problem I have not been able to fix while upgrading the package to Julia v1 is related to `ltriang` function. Actually, the problem is triggered by one of the calls to `gcrd` function in test and I traced it down to `ltriang` (in fact, `_ltriang`).

Preparation of the data first:
```julia
using Polynomials, PolynomialMatrices
s = variable("s")
M = [s^2 0 1; 0 s^2 1+s]
P = PolyMatrix(M)
```
Now comes the problem:
```julia
julia> L, U = ltriang(P)
ERROR: MethodError: Cannot `convert` an object of type Nothing to an object of type Int64
Closest candidates are:
  convert(::Type{T<:Number}, ::T<:Number) where T<:Number at number.jl:6
  convert(::Type{T<:Number}, ::Number) where T<:Number at number.jl:7
  convert(::Type{T<:Integer}, ::Ptr) where T<:Integer at pointer.jl:23
  ...
Stacktrace:
 [1] setindex!(::Array{Int64,1}, ::Nothing, ::Int64) at ./array.jl:766
 [2] _ltriang(::PolyMatrix{Float64,Array{Float64,2},Val{:s},2}, ::Bool, ::Int64) at /home/hurak/.julia/packages/PolynomialMatrices/z1iRu/src/reductions.jl:262
 [3] _ltriang(::PolyMatrix{Float64,Array{Float64,2},Val{:s},2}, ::Bool, ::Int64) at /home/hurak/.julia/packages/PolynomialMatrices/z1iRu/src/reductions.jl:318 (repeats 2 times)
 [4] ltriang(::PolyMatrix{Float64,Array{Float64,2},Val{:s},2}) at /home/hurak/.julia/packages/PolynomialMatrices/z1iRu/src/reductions.jl:178
 [5] top-level scope at none:0
```
Calling the function with the fixed degree `du` and iterations switched off gives the same error:

```julia
julia> iterative = false
false

julia> dᵤ = 2
2

julia> L, U = ltriang(P,iterative,dᵤ)
ERROR: MethodError: Cannot `convert` an object of type Nothing to an object of type Int64
Closest candidates are:
  convert(::Type{T<:Number}, ::T<:Number) where T<:Number at number.jl:6
  convert(::Type{T<:Number}, ::Number) where T<:Number at number.jl:7
  convert(::Type{T<:Integer}, ::Ptr) where T<:Integer at pointer.jl:23
  ...
Stacktrace:
 [1] setindex!(::Array{Int64,1}, ::Nothing, ::Int64) at ./array.jl:766
 [2] _ltriang(::PolyMatrix{Float64,Array{Float64,2},Val{:s},2}, ::Bool, ::Int64) at /home/hurak/.julia/packages/PolynomialMatrices/z1iRu/src/reductions.jl:262
 [3] ltriang(::PolyMatrix{Float64,Array{Float64,2},Val{:s},2}, ::Bool, ::Int64) at /home/hurak/.julia/packages/PolynomialMatrices/z1iRu/src/reductions.jl:178
 [4] top-level scope at none:0
```
In addition, it is quite confusing that sometimes after calling the function `ltriang` a few times, switching between the two ways, the function occassionally finishes without an error:

```julia
julia> julia> L, U = ltriang(P,iterative,dᵤ)
(Poly{Float64}[Poly(-1.0) Poly(0.0) Poly(0.0); Poly(-1.0 - 1.0*s) Poly(-1.1547005383792515*s^2) Poly(0.0)], Poly{Float64}[Poly(0.0) Poly(0.28867513459481287) Poly(-0.5000000000000001); Poly(0.0) Poly(-0.8660254037844386 + 0.2886751345948128*s) Poly(-0.49999999999999994 - 0.5*s); Poly(-1.0) Poly(-0.28867513459481287*s^2) Poly(0.5*s^2)])
```
The returned solution is indeed correct:
```julia
julia> P*U-L
2×3 PolyMatrix{Float64,Array{Float64,2},Val{:s},2}:
 Poly(0.0)  Poly(0.0)  Poly(0.0)
 Poly(0.0)  Poly(0.0)  Poly(0.0)
```
The problem only appers with a wide polynomial matrix. With a square matrix is works correctly.

**Comment by neveritt** [comment]
It is a very strange behavior you have observed!

I am still trying to understand the new Julia. I have found one instance in
multiplication of to PolynomialMatrices where it was assumed that the
function
insert!, made a copy, however it was not. I have tried to look over the
code for more instances of this bug, but could not find any.

In this case, hopefully your observation that it sometimes finishes with
the correct result can give some insight into what might be wrong.

I was hoping we could fix this error and the bug in colred/rowred and then
make a new release.
Then I think the package would be in pretty good shape and we could move it
to JuliaPolynomialMatrices. what do you think?

Best wishes,
Niklas

Den lör 7 sep. 2019 kl 15:49 skrev Zdeněk Hurák <notifications@github.com>:

**Comment by hurak** [comment]
> It is a very strange behavior you have observed! [...] hopefully your observation that it sometimes finishes with the correct result can give some insight into what might be wrong. 

I suspected for a moment that this might be related to the issue of viewing vs copying vs deep-copying the input objects. One of the changes in Julia from 0.6 to 1.0 is that `A.'` is no longer used for transposition of an array/matrix `A` and instead `transpose(A)` or `copy(transpose(A))` is recommended (note that the change is not even mentioned in [the release notes for 0.7](https://docs.julialang.org/en/v0.7/NEWS/) and instead is only mentioned in warning messages in 0.7). I applied this change to the `PolynomialMatrix.jl` code in all the places that I could find. But observing the behavior of some calls of the function failing while some other giving a correct result), I started suspecting if the related function(s) are somehow overrwriting the input data. But I checked and it was not the case. Still I wanted to share this suspicion, maybe I am still missing something. 

> I was hoping we could fix this error and the bug in colred/rowred and then make a new release. Then I think the package would be in pretty good shape and we could move it to JuliaPolynomialMatrices. what do you think? Best wishes, Niklas 

Indeed, I can tell you I spent already a few hours with the `ltriang` problem exactly with the hope that once it is fixed, the package would be ready for a new release. If you are willing to do so, the package can then be moved to the `JuliaPolynomialMatrices` organization. Let's now work a bit on fixing the problems first. Regards and thanks.
